README.md
NewMedia! Denver's nmdbase cookbook
=============================

nmdbase (1.0.9) Manages ldap client, yubico pam, ssl certificates and unattended updates.

This is a base cookbook for all NewMedia Denver servers. It contains core functionality necessary for standardized integration into our broader systems. In the spirit of open source, we are going to illustrate how to properly craft, and deliver, fantastically reliable and secure infrastructure.We use this recipe to enable two factor authentication for ssh accounts. The first factor is a plain text password the user knows. The second is a YubiKey usb hardware device. The instance is configured to create a new linux account on the machine if both factors authenticate. We also use this recipe to install fail2ban to protect against repeated ssh failures and ssh ddos attacks. The final task performed by this recipe is to enable the instance as a chef client so that it is regularly checking in with our chef servers. Test kitchen is configured to expect that the environment variable DATA_BAGS_PATH be set.  To use the example databags set DATA_BAGS_PATH to test/integration/data_bags/ ie export DATA_BAGS_PATH=test/integration/data_bags/ and set use_encrypted_databags to :no.

Requirements
------------

### Platforms

`ubuntu >= 14.04`

`centos >= 6.0`

### Dependencies

`chef-client >= 0.0.0`

`apt >= 0.0.0`

`openssh >= 0.0.0`

`fail2ban >= 0.0.0`

`simple_iptables >= 0.0.0`

`logwatch >= 0.0.0`

`postfix >= 0.0.0`


Attributes
----------


Recipes
-------

    nmdbase::default
      Enables the chef-client service on a schedule in addition to each of the other recipes in this cookbook.

    nmdbase::ldap
      Installs and configures ldap pam authentication.

    nmdbase::ssl
      Manages the organization specific ssl certificates.

    nmdbase::yubico
      Installs and configures yubico pam authentication.

    nmdbase::iptables
      Configures iptables.  Uses the recipe simple_iptables https://github.com/rtkwlf/cookbook-simple-iptables  to manage rules / polices.  example:  "nmdbase": {           "simple_iptables_policy": {             "policy 1": {               "name": "INPUT",               "table": "nat",               "defined_policy": "ACCEPT"             }           },           "simple_iptables_rules": {             "rule 1": {               "name": "icmp",               "chain": "INPUT",               "rule": "--proto icmp",               "jump": "ACCEPT",               "weight": 2             }           }         }

Testing and Utility
-------
    <Rake::Task default => [style, integration:kitchen:all]>

    <Rake::Task integration:kitchen:all => [default-ubuntu-1404-vmware, default-ubuntu-1404-virtualbox, default-centos-65-vmware, default-centos-65-virtualbox, ldap-ubuntu-1404-vmware, ldap-ubuntu-1404-virtualbox, ldap-centos-65-vmware, ldap-centos-65-virtualbox, yubico-ubuntu-1404-vmware, yubico-ubuntu-1404-virtualbox, yubico-centos-65-vmware, yubico-centos-65-virtualbox, yubico-ldap-ubuntu-1404-vmware, yubico-ldap-ubuntu-1404-virtualbox, yubico-ldap-centos-65-vmware, yubico-ldap-centos-65-virtualbox]>
      Run all test instances

    <Rake::Task integration:kitchen:default-centos-65-virtualbox => []>
      Run default-centos-65-virtualbox test instance

    <Rake::Task integration:kitchen:default-centos-65-vmware => []>
      Run default-centos-65-vmware test instance

    <Rake::Task integration:kitchen:default-ubuntu-1404-virtualbox => []>
      Run default-ubuntu-1404-virtualbox test instance

    <Rake::Task integration:kitchen:default-ubuntu-1404-vmware => []>
      Run default-ubuntu-1404-vmware test instance

    <Rake::Task integration:kitchen:ldap-centos-65-virtualbox => []>
      Run ldap-centos-65-virtualbox test instance

    <Rake::Task integration:kitchen:ldap-centos-65-vmware => []>
      Run ldap-centos-65-vmware test instance

    <Rake::Task integration:kitchen:ldap-ubuntu-1404-virtualbox => []>
      Run ldap-ubuntu-1404-virtualbox test instance

    <Rake::Task integration:kitchen:ldap-ubuntu-1404-vmware => []>
      Run ldap-ubuntu-1404-vmware test instance

    <Rake::Task integration:kitchen:yubico-centos-65-virtualbox => []>
      Run yubico-centos-65-virtualbox test instance

    <Rake::Task integration:kitchen:yubico-centos-65-vmware => []>
      Run yubico-centos-65-vmware test instance

    <Rake::Task integration:kitchen:yubico-ldap-centos-65-virtualbox => []>
      Run yubico-ldap-centos-65-virtualbox test instance

    <Rake::Task integration:kitchen:yubico-ldap-centos-65-vmware => []>
      Run yubico-ldap-centos-65-vmware test instance

    <Rake::Task integration:kitchen:yubico-ldap-ubuntu-1404-virtualbox => []>
      Run yubico-ldap-ubuntu-1404-virtualbox test instance

    <Rake::Task integration:kitchen:yubico-ldap-ubuntu-1404-vmware => []>
      Run yubico-ldap-ubuntu-1404-vmware test instance

    <Rake::Task integration:kitchen:yubico-ubuntu-1404-virtualbox => []>
      Run yubico-ubuntu-1404-virtualbox test instance

    <Rake::Task integration:kitchen:yubico-ubuntu-1404-vmware => []>
      Run yubico-ubuntu-1404-vmware test instance

    <Rake::Task readme => [utility:readme]>
      Generate README.md

    <Rake::Task spec => [style:spec]>
      Run rspec tests.

    <Rake::Task style => [style:chef, style:ruby]>
      Run all style checks

    <Rake::Task style:chef => []>
      Run Chef style checks

    <Rake::Task style:ruby => []>
      Run Ruby style checks

    <Rake::Task style:ruby:auto_correct => []>
      Auto-correct RuboCop offenses

    <Rake::Task style:spec => []>
      Run ChefSpec examples. Specify OS to test either with rake "spec[rhel]" (Redhat,centos etc) or rake "spec[ubuntu]". Defaults to both

    <Rake::Task travis => [style]>
      Run all tests on Travis

    <Rake::Task utility:readme => []>
      Generate the Readme.md file.

License and Authors
------------------

The following engineers have contributed to this code:
 * [Kevin Bridges](https://github.com/cyberswat) - 151 commits
 * [David Arnold, DavidXArnold](https://github.com/DavidXArnold) - 95 commits
 * [Making GitHub Delicious.](https://github.com/waffle-iron) - 1 commits
 * [Brandon Williams](https://github.com/bw411) - 9 commits

Copyright:: 2014, NewMedia Denver

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Contributing
------------

We welcome contributed improvements and bug fixes via the usual workflow:

1. Fork this repository
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new pull request


CHANGELOG.md
nmdbase CHANGELOG
====================

This file is used to list changes made in each version of the nmdbase
cookbook.

1.0.10 13-August-2014
-----
- [David Arnold] - Set iptables to reload vs restart.
1.0.9 14-July-2014
-----
- [David Arnold] - Added logwatch & postfix recipe to default and updated berksfile GEM version.
- 
1.0.8 7-July-2014
-----
- [David Arnold] - Changed rhel default nsswitch attribute.

1.0.7 1-July-2014
-----
- [David Arnold] - Added iptables recipe.

1.0.6 27-June-2014
-----
- [David Arnold] - Changed data bag encryption attribute to string vs symbol.

1.0.5 24-June-2014
-----
- [David Arnold] - Changed data bag encryption level to default to :no.

1.0.4 24-June-2014
-----
- [David Arnold] - Updated to support encrypted data bags.

1.0.3 19-June-2014
-----
- [David Arnold] - Initial release of nmdbase that includes ldap, sssd and yubico config.

- - -

# 0.1.0 07-April-2014

* Initial Release
  -  nmdbase::default
    -  Includes the fail2ban recipe.
    -  Configures this instance as an LDAP client.
    -  Configures this instance as an Yubico API client.
    -  Configures this instance as a chef client.
    -  Configures the chef-client service.
  -  nmdbase::ldap
    -  Installs the LDAP package to set this instance up as a client.
    -  Installs LDAP command line utilities.
    -  Configures the LDAP connection for this client.
    -  Installs the LDAP secret authentication content.
    -  Modifies the Name Service Switch to use LDAP.
    -  It configures the PAM common session to create users from LDAP.
  -  nmdbase::yubico
    -  Includes the openssh recipe.
    -  Installs the openssh-client.
    -  Installs the openssh-server.
    -  Enables the ssh service.
    -  Starts the ssh service.
    -  Creates the ssh configuration.
    -  Creates the sshd configuration.
    -  Installs python-software-properties.
    -  Adds the yubico repositories.
    -  Notifies apt-get update when adding yubico repositories.
    -  Installs the libpam-yubico package from the yubico repositories.
    -  Creates a global yubico auth file.
    -  Configures the sshd PAM module.
    -  Prepares for PAM debug logging.



metadata.rb
# encoding: utf-8
name 'nmdbase'
maintainer 'NewMedia! Denver'
maintainer_email 'support@newmediadenver.com'
license 'Apache 2.0'

version '1.0.10'
supports 'ubuntu', '>= 14.04'
supports 'centos', '>= 6.0'

desc = 'Manages ldap client, yubico pam, ssl certificates and unattended '
desc << 'updates.'
description desc

desc = 'Acts as a base recipe for chef-client and security components '
desc << '(ldap/yubikey two factor authentication).'

desc = 'This is a base cookbook for all NewMedia Denver servers. It contains '
desc << 'core functionality necessary for standardized integration into our '
desc << 'broader systems. In the spirit of open source, we are going to '
desc << 'illustrate how to properly craft, and deliver, fantastically '
desc << 'reliable and secure infrastructure.'

desc << 'We use this recipe to enable two factor authentication for ssh '
desc << 'accounts. The first factor is a plain text password the user knows. '
desc << 'The second is a YubiKey usb hardware device. The instance is '
desc << 'configured to create a new linux account on the machine if both '
desc << 'factors authenticate. We also use this recipe to install fail2ban to '
desc << 'protect against repeated ssh failures and ssh ddos attacks. The '
desc << 'final task performed by this recipe is to enable the instance as a '
desc << 'chef client so that it is regularly checking in with our chef '
desc << 'servers. Test kitchen is configured to expect that the environment'
desc << ' variable DATA_BAGS_PATH be set.  To use the example databags '
desc << 'set DATA_BAGS_PATH to test/integration/data_bags/ ie'
desc << ' export DATA_BAGS_PATH=test/integration/data_bags/ and set '
desc << 'use_encrypted_databags to :no.'
long_description desc

desc = 'Enables the chef-client service on a schedule in addition to each of '
desc << 'the other recipes in this cookbook.'
recipe 'nmdbase::default', desc

provides 'nmdbase::ldap'
recipe 'nmdbase::ldap', 'Installs and configures ldap pam authentication.'

provides 'nmdbase::ssl'
recipe 'nmdbase::ssl', 'Manages the organization specific ssl certificates.'

provides 'nmdbase::yubico'
recipe 'nmdbase::yubico', 'Installs and configures yubico pam authentication.'

provides 'nmdbase::iptables'
desc = 'Configures iptables.  Uses the recipe '
desc << 'simple_iptables https://github.com/rtkwlf/cookbook-simple-iptables '
desc << ' to manage rules / polices. '
desc << ' example: '
desc << ' "nmdbase": {'
desc << '           "simple_iptables_policy": {'
desc << '             "policy 1": {'
desc << '               "name": "INPUT",'
desc << '               "table": "nat",'
desc << '               "defined_policy": "ACCEPT"'
desc << '             }'
desc << '           },'
desc << '           "simple_iptables_rules": {'
desc << '             "rule 1": {'
desc << '               "name": "icmp",'
desc << '               "chain": "INPUT",'
desc << '               "rule": "--proto icmp",'
desc << '               "jump": "ACCEPT",'
desc << '               "weight": 2'
desc << '             }'
desc << '           }'
desc << '         }'

recipe 'nmdbase::iptables', desc

depends 'chef-client'
depends 'apt'
depends 'openssh'
depends 'fail2ban'
depends 'simple_iptables'
depends 'logwatch'
depends 'postfix'
